pkg/addons2/upgrade.go

```diff
@@ -123,31 +122,16 @@ func upgradeAddOn(ctx context.Context, hcli helm.Client, kcli client.Client, in
 		return errors.Wrap(err, "failed to set condition status")
 	}
 
-	defer func() {
-		if r := recover(); r != nil {
-			finalErr = fmt.Errorf("upgrading %s recovered from panic: %v: %s", addon.Name(), r, string(debug.Stack()))
-		}
-
-		status := metav1.ConditionTrue
-		reason := "Upgraded"
-		message := ""
-
-		if finalErr != nil {
-			status = metav1.ConditionFalse
-			reason = "UpgradeFailed"
-			message = helpers.CleanErrorMessage(finalErr)
-		}
-
-		if err := setCondition(ctx, kcli, in, conditionName(addon), status, reason, message); err != nil {
-			slog.Error("Failed to set condition status", "error", err)
-		}
-	}()
-
 	// TODO (@salah): add support for end user overrides
 	overrides := addOnOverrides(addon, in.Spec.Config, nil)
 
-	if err := addon.Upgrade(ctx, kcli, hcli, overrides); err != nil {
-		return errors.Wrap(err, addon.Name())
+	upgradeErr := addon.Upgrade(ctx, kcli, hcli, overrides)
+	if err := setFinalCondition(ctx, kcli, in, addon, upgradeErr); err != nil {
```

Comment:
it seems like an anti-pattern to not check the err immediately, no? why not do what you did in the v2 migrate code? would probably have to break `setFinalCondition` into two functions, but i think that's fine?